### PR TITLE
session: reset CTEStorageMap before txn conflict and retry()

### LINKED ISSUE #46522: CTE query execute failed: type assertion for CTEStorageMap failed
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

run cte-query

### 2. What did you expect to see? (Required)
run success and no error

### 3. What did you see instead (Required)
run failed and stack info

### 4. What is your TiDB version? (Required)

v6.5.3

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #46522

Problem Summary:

`sctx.CTEStorageMap` was set to `nil` in finishStmt, the code may call `retry()` in `doCommitWithRetry()`  and rebuild the statement, during the rebuild process, `sctx.CTEStorageMap` type assertion fail.

### Check List

Tests 

- [ ] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by jackysp — DISMISSED** [review]
LGTM

File: tests/integrationtest/t/executor/cte.test [hunk 1]
```diff
@@ -153,3 +153,19 @@ create table t1(c1 int);
...
 with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 c1 from cte1 limit 1 offset 100) select * from cte1;
 set tidb_max_chunk_size=default;
+
+# TestIssue46522
```
Comment (jackysp) [inline tests/integrationtest/t/executor/cte.test:157]: I cannot reproduce the original panic in the issue with this ticket.

**Review by bb7133 — APPROVED** [review]
LGTM

File: pkg/session/session.go [hunk 2]
```diff
@@ -1146,6 +1146,7 @@ func (s *session) retry(ctx context.Context, maxCnt uint) (err error) {
 		for i, sr := range nh.history {
 			st := sr.st
 			s.sessionVars.StmtCtx = sr.stmtCtx
+			s.sessionVars.StmtCtx.CTEStorageMap = map[int]*executor.CTEStorages{}
```
Comment (bb7133) [inline pkg/session/session.go:1149]: I believe this missing code is a typical example of 'session context' issue of TiDB. /cc @lcwangchao

Comment (lcwangchao) [reply pkg/session/session.go:1149]: But I think the actual reason is the code path in retry is different with executing a new statement normally. In normal cases, `ResetContextOfStmt` is called when we enter a new statement, but in retry logic, we use `ResetForRetry` instead.

Comment (jackysp) [reply pkg/session/session.go:1149]: Yes, but we cannot totally reset the stmtctx for retry.

File: tests/integrationtest/t/executor/cte.test (hunk 1 above)
Comment (tiancaiamao) [reply tests/integrationtest/t/executor/cte.test:157]: image

This is the error stack, 

```
[err="type assertion for CTEStorageMap failed
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).buildCTE
	/home/genius/project/src/github.com/pingcap/tidb/pkg/executor/builder.go:5277
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).build
	/home/genius/project/src/github.com/pingcap/tidb/pkg/executor/builder.go:308
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).buildInsert
	/home/genius/project/src/github.com/pingcap/tidb/pkg/executor/builder.go:976
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).build
	/home/genius/project/src/github.com/pingcap/tidb/pkg/executor/builder.go:194
github.com/pingcap/tidb/pkg/executor.(*ExecStmt).buildExecutor
	/home/genius/project/src/github.com/pingcap/tidb/pkg/executor/adapter.go:1203
github.com/pingcap/tidb/pkg/executor.(*ExecStmt).Exec
	/home/genius/project/src/github.com/pingcap/tidb/pkg/executor/adapter.go:550
github.com/pingcap/tidb/pkg/session.(*session).retry
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/session.go:1177
github.com/pingcap/tidb/pkg/session.(*session).doCommitWithRetry
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/session.go:898
github.com/pingcap/tidb/pkg/session.(*session).CommitTxn
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/session.go:1001
github.com/pingcap/tidb/pkg/session.autoCommitAfterStmt
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/tidb.go:298
github.com/pingcap/tidb/pkg/session.finishStmt
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/tidb.go:260
github.com/pingcap/tidb/pkg/session.runStmt
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/session.go:2413
github.com/pingcap/tidb/pkg/session.(*session).ExecuteStmt
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/session.go:2236
github.com/pingcap/tidb/pkg/server.(*TiDBContext).ExecuteStmt
	/home/genius/project/src/github.com/pingcap/tidb/pkg/server/driver_tidb.go:293
github.com/pingcap/tidb/pkg/server.(*clientConn).handleStmt
	/home/genius/project/src/github.com/pingcap/tidb/pkg/server/conn.go:2045
github.com/pingcap/tidb/pkg/server.(*clientConn).handleQuery
	/home/genius/project/src/github.com/pingcap/tidb/pkg/server/conn.go:1835
github.com/pingcap/tidb/pkg/server.(*clientConn).dispatch
	/home/genius/project/src/github.com/pingcap/tidb/pkg/server/conn.go:1322
github.com/pingcap/tidb/pkg/server.(*clientConn).Run
	/home/genius/project/src/github.com/pingcap/tidb/pkg/server/conn.go:1101
github.com/pingcap/tidb/pkg/server.(*Server).onConn
	/home/genius/project/src/github.com/pingcap/tidb/pkg/server/server.go:701
runtime.goexit
	/home/genius/project/go/src/runtime/asm_amd64.s:1650
previous statement: insert into issue46522 with t1 as (select id+1 from  issue46522 where id = 1) select * from t1"]
```

I think they are the same type. @jackysp 
doCommitWithRetry() -> retry() -> ... -> buildCTE() -> type assertion for CTEStorageMap failed

File: pkg/session/session.go (hunk 2 above)
Comment (tiancaiamao) [reply pkg/session/session.go:1149]: The retry behavior is disabled by default.
It cause lost update https://github.com/pingcap/tidb/issues/10075

Comment (tiancaiamao) [reply pkg/session/session.go:1149]: I really hope one day we can get rid of the legacy code like this.
We can simplify the logic a lot!

**Pull request merged**